### ISSUE TITLE
Handling of Core & plugins with git subtrees and Composer packages

### ISSUE THREAD

**Issue by mnapoli** [piwik/piwik#6757]
Goal: find a solution for:

- #6605 Eliminate git submodules
- #4485 Support "requiring" other plugins from one plugin
- #5921 Rename `plugin.json` to `composer.json`

We could use git subtrees and sub-repositories, for example:

- `/` -> git repository & GitHub project (`piwik/piwik`) & composer package (`piwik/piwik`)
    - `core/` -> git subtree & GitHub project (`piwik/core`) & composer package (`piwik/core`)
    - `plugins/`
        - `CustomAlerts/` -> git subtree & GitHub project (`piwik/plugin-CustomAlerts`) & composer package (`piwik/plugin-CustomAlerts`)
        - ...

Regarding Composer, we would then use the [`replace`](https://getcomposer.org/doc/04-schema.md#replace) feature, e.g.:

- `piwik/piwik` replaces `piwik/core`, `piwik/plugin-CustomAlerts`, …
- plugins require `piwik/core` (and **not** `piwik/piwik`), and maybe any other plugin: that way we can also run the tests for plugins independently from Piwik by just using Composer (which will install `piwik/core` and not the whole `piwik/piwik`)

Maybe all plugins could become subtrees? (to have their `composer.json`). That would mean a lot of repositories though.

This is what Symfony does so that `symfony/symfony` contains the whole framework, and then `symfony/console` is just the console component. Have a look at the [`composer.json`](https://github.com/symfony/symfony/blob/2.7/composer.json#L24-L69). And as you can see, [`symfony/Console`](https://github.com/symfony/Console) is just a read-only git subtree.

This is also what Laravel does, for example: [illuminate/events](https://github.com/illuminate/events).

Fabien Potencier mentioned in [this article](http://symfony.com/blog/symfony2-components-as-standalone-packages) that he is using a script to maintain all subtree in sync with the main one, I'm guessing Laravel do the same. Maybe we could too.

Pros:

- Piwik Core & plugins integrated in the main repo would have the same versioning as Piwik (which makes everything easier: same versions, same branches, same tags, etc...)
- Each plugin could have a `composer.json` and be a Composer package
- Piwik Core could be a Composer package by itself (allowing to be required by plugins, which would let us run plugin tests independently of the whole Piwik application, I like that)

*What about the UI tests?* I don't think this problem would be solved.

---

**TL/DR:** Do like Symfony with their components & bundles:

- some are in their own separate repositories
- some are in the main repository but with a *git subtree*, which lets them also be their own repository and composer package

**Comment by mattab** [comment]
Drupal is about to git-tree their repo as well for some reason https://www.drupal.org/node/2352091#comment-9378633

**Comment by mnapoli** [comment]
I'm closing this issue because it mixes Composer and "getting rid of submodules".

The "getting rid of submodules" is discussed in #6605 (using git subtrees).
